doc/memos/rdm-draft-petry-riot-design-goals.md

```diff
@@ -0,0 +1,345 @@
+
+## 2.6. Smart home devices
+
+Easy to use devices can let consumers monitor and control their home
+automatically. The devices need to be able to:
+
+  - Sense temperature, light, humidity, smoke, etc
+  - Integrate with home appliances to control temperature, light, power usage,
+    etc
+  - Be usable but secure for people with no technical knowledge 
+  - Connect to a commercial home gateway 
+  - Communicate over common home IoT protocols such as ZigBee, Bluetooth, WiFi,
+    etc
+  - Connect to the building's power or provide their own power, depending on
+    the product
+  
+## 2.7. Daughterboards
+
+Plug-in boards can give devices immediate support for a protocol or standard,
+or let them outsource a task from the main processor. This requires the board
+to:
+
+  - Perform its promised task reliably
+  - Integrate and communicate reliably with the main processor
+  - Include and support any relevant hardware
+
+## 2.8. Education
+
+The broad technical scope of RIOT makes it useful as a basis for education.
+This requires:
+
+  - The presence of didactic materials related to RIOT
+  - Clear and easy to find documentation
+  - A general availability of supported hardware and tools
+
+# 3. Design philosophies
+
+RIOT satisfies the requirements of all the use cases given above, and does so
+in its own unique way, a way that reflects the collective personality of the
+RIOT community. RIOT is the Friendly IoT Operating System: it is friendly to
+users, allowing them to easily build and deploy IoT solutions that are stable
+and trouble-free.
+
+Below are the design philosophies that are typically followed by developers to
+achieve this, including descriptions of tradeoffs between them and where the
+resolutions typically fall.
+
+## Suitability for constrained devices
+
+"Constrained" means that available memory, energy, and processor cycles are so
+reduced as to become a dominant consideration in design requirements [1]. 
+
+#### Energy efficiency
+
+RIOT nodes sometimes need to last for several years without external power, so
+they need to manage energy carefully. RIOT's tickless scheduler lets devices
+sleep while they aren't collecting data.
+
+Modules outside the core should leverage the benefits and address the
+programming challenges of such a scheduler. They shouldn't demand that users do
+the same.  They should, however, allow users to manage power through different
+modes and functions.
+
+#### Small memory footprint
+
+Most of RIOT's targeted use cases are well addressed by devices in class 1 of
+the taxonomy presented in [1]. If small price differences are important or the
+energy budget is particularly tight, the available memory might be near the
+bottom of this class. Over-the-air updating currently reduces the available ROM
+by over half.
+
+RIOT should provide out-of-the-box support for devices with ~10 KiB of
+available RAM and ~100 KiB of ROM. It should be just as possible to address
```

Comment:
I would propose to remove the remaining two sentences (starting with "It should be just as possible..."). Below class 1 is not focus and - also in the light of the previous discussion - maybe considered an over-selling.

Reply:
Well, to me this discussion is artificial and a statement  "with manual optimizations and specialised modules" IMO does not serve well as guidance for developing a versatile, multi-purpose OS.  
I would still propose to remove at least the last sentence - "Users should be given the choice of what they want to spend their memory on." - because this is an empty phrase given that a desired function comes at a minimal cost.

Reply:
The sentiment that I've picked up is that this is a real discussion - for example, people have cited the comments of one of the speakers at the last Summit that our key differentiator is that we have a slimmer memory footprint than the competition and that we should continue to strengthen that aspect. That to me is not at all artificial (if I understand correctly what you meant by the word).  

Re the language - to make it more tight - perhaps "with manual optimizations and specialized modules" -> "via configuration and modules that prioritize memory efficiency in their design, without of course departing from specifications". And delete the last sentence.

Reply:
+1 for removing ""Users should be given the choice of what they want to spend their memory on."

Other than that, I don't understand the aversion to supporting low-end embedded devices.

To me RIOT is the first choice for creating *any* application involving MCUs. I do choose it even on the smallest supported platforms (e.g., Arduino Uno, with 2kB of RAM), e.g., for simple Arduino-style control tasks. For reference, hello-world on that board uses ~1250B RAM, so there's room to spare for quite some control logic.
While on a board that small I can certainly not enjoy RIOT's awesome networking options, many of RIOT's perks make it a valid first choice (even compared to the Arduino libraries):

- nice and compact periph APIs
- multi-threading
- portability (upgrade path) to other (larger) MCUs using the same application code
- large collection of supported sensors (with similar, familiar APIs)

I don't see why using RIOT on those systems is only possible "with manual optimizations and specialized modules".

It all depends on the application, but RIOT is a good fit for ~2kB RAM systems for use in control tasks (which means basically anything that the supported Atmega 328p is capable of). Why state otherwise in the design goals?

Reply:
There is no aversion against small memory footprint or the like - if RIOT achieves highest efficiency that's great. Still I'm hesitant with raising the wrong expectations - and manually optimized specialized modules sound like bare-metal firmware to me. 
Maybe a good solution here is to state that the RIOT kernel consumes 1.5kB RAM and - depending on the application and the modules loaded - the overall size of the system can be below 10kB.

Reply:
> and manually optimized specialized modules sound like bare-metal firmware to me.

@kaspar030 mentioned no such thing about this. Fact is we do support small devices (without networking capabilities) like the Arduino Uno and I agree with @kaspar030 that for some developers RIOT might be the nicer API than the somewhat clunky, C++-based Arduino API.

Reply:
> IMO it is important to be precise here so not to raise the impression that RIOT (including all prominent features) is a thing that can run usually on 2 kB RAM.

Well, isn't it important to raise awareness that for the 2kB Arduino class, RIOT is actually a good choice, for the type of applications that can usually be done on those devices?

Reply:
I don't object mentioning, as long as the case is clearly stated. My fear is that people get disappointed if expectations are raised that later turn out to be unrealistic.

Reply:
If the size of the kernel is now relatively stable and at a size we're happy with, it could be helpful to state the size along with a statement that it shouldn't grow more than this size in its default configuration. We could also be more specific about the kinds of applications that can be run on devices with 2KiB RAM and 10 KiB ROM? Some of the aspects mentioned:
- no networking support
- support for sensors and actuators with straightforward (not much control logic in the driver, for example) functionality
- space for simple control and sensing application logic
If anyone can think of any more detail or refine these statements further, that would be useful

Reply:
I think we're looking at this from the wrong angle. We're trying to make "there are valid uses cases on 2kB AVR" fit with "our main focus is IoT on class 1 upwards" thinking. I'm not sure we can converge on a "primary use case" as in "class 1 IoT", which I personally just consider as one of the many uses of RIOT. So how about we explain that RIOT's memory usage *depends* and give examples of actual applications:

How about something like this:

```
Apart from being optimized for low memory usage, RIOT tries very hard to be as modular as possible so unused features don't use up precious RAM or flash. Almost all features are provided as optional modules that have to be enabled explicitly at compile time.
A minimal RIOT configuration starts at around <2KiB flash and <1.5KiB RAM (including stack space for one thread and ISRs).

Starting from there, the memory usage depends on the enabled features:

- non-networked control loop / sensing applications can fit on very small MCUs (eg., an Atmega328P with 2KiB RAM)
- 6lowPAN networking currently starts at ~10KiB RAM
- a file system adds X kB
- Javascript needs Y kIB
- security, OTA  Z KiB (might want to link the coming OTA paper for numbers)
- use of C library (it's huge) vs use of more optimized libs (printf vs. fmt)
- a 6lowpan/RPL enabled CoAP server with secure OTA enabled needs xx KiB RAM, yy KiB flash
```

Reply:
@kaspar030 I agree on the 'angle' and also on the approach of just listing relevant configurations. This will be (a) informative and (b) transparent, so the reader can judge what is important for his or her context.

Reply:
I think being more specific is a good path. Would this be able to be directed further in a way which helps a RIOT kernel developer to judge how far they should optimize their new module for size? For example, we could angle it as a list of indicative sizes (+ size range, special cases, etc) for certain types of feature? And point out making available optimized libraries alongside more fully functional ones?

Reply:
Also if anyone has any values to propose for X, Y, Z which they know off the top of their head, that would be great, otherwise I'll get some build size metrics and propose some values myself